src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -249,26 +250,70 @@ public TagHelperContent EndTagHelperWritingScope()
             return tagHelperContentWrapperTextWriter.Content;
         }
 
-        /// <summary>
-        /// Writes an <see cref="ITextWriterCopyable"/> to the <see cref="Output"/>.
-        /// </summary>
-        /// <param name="copyableTextWriter">Contains the data to be written.</param>
-        public void Write(ITextWriterCopyable copyableTextWriter)
+        public void WriteTagHelper(TagHelperExecutionContext tagHelperExecutionContext)
         {
-            WriteTo(Output, copyableTextWriter);
+            WriteTagHelperTo(Output, tagHelperExecutionContext);
         }
 
-        /// <summary>
-        /// Writes an <see cref="ITextWriterCopyable"/> to the <paramref name="writer"/>.
-        /// </summary>
-        /// <param name="writer">The <see cref="TextWriter"/> to which the
-        /// <paramref name="copyableTextWriter"/> is written.</param>
-        /// <param name="copyableTextWriter">Contains the data to be written.</param>
-        public void WriteTo([NotNull] TextWriter writer, ITextWriterCopyable copyableTextWriter)
+        public void WriteTagHelperTo([NotNull] TextWriter writer, TagHelperExecutionContext tagHelperExecutionContext)
         {
-            if (copyableTextWriter != null)
+            var tagHelperOutput = tagHelperExecutionContext.Output;
+            var isTagNameNullOrWhitespace = string.IsNullOrWhiteSpace(tagHelperOutput.TagName);
+
+            if (!isTagNameNullOrWhitespace)
             {
-                copyableTextWriter.CopyTo(writer);
+                writer.Write('<');
+                writer.Write(tagHelperOutput.TagName);
+
+                foreach (var attribute in tagHelperOutput.Attributes)
+                {
+                    var value = HtmlEncoder.HtmlEncode(attribute.Value);
+                    writer.Write(' ');
+                    writer.Write(attribute.Key);
+                    writer.Write("=\"");
+                    writer.Write(value);
+                    writer.Write('"');
+                }
+
+                if (tagHelperOutput.SelfClosing)
+                {
+                    writer.Write(" /");
+                }
+
+                writer.Write('>');
+            }
+
+            if (isTagNameNullOrWhitespace || !tagHelperOutput.SelfClosing)
+            {
+                WriteTagHelperContentTo(writer, tagHelperOutput.PreContent);
+                if (tagHelperOutput.IsContentModified)
+                {
+                    WriteTagHelperContentTo(writer, tagHelperOutput.Content);
+                }
+                else if (tagHelperExecutionContext.ChildContentRetrieved)
+                {
+                    WriteTagHelperContentTo(writer, tagHelperExecutionContext.GetChildContentAsync().Result);
```

Comment:
@NTaylorMullen why would that change be part of this PR?  the generated code this replaces did the same things.  so there's no functional difference (yet).  agree we should clean things up and use `Write*Async(...).Wait()` or (if we ever can) `await Write*Async()` but that's a larger change that should take advantage of `TextWriter.WriteAsync()` as well.